### PR TITLE
dnsdist: Add an API endpoint to remove entries from caches

### LINKED ISSUE #6154: dnsdist: Implement cache expungeByName support in the APIs
- Program: dnsdist
 - Issue type: Feature request

### Short description

Currently, the dnsdist APIs do not expose any endpoint to manipulate the records cache (e.g. evict some records). 
This issue aims to request the evaluation of the implementation an endpoint to manipulate via APIs dnsdist's per-pools records cache.

### Usecase & Description

Deploying dnsdist as additional cache/protection layer in front of Authoritative DNS servers require specifically maintenance operations in order to make sure to evict dnsdist records cache everytime a DNS record is updated on the backend Authoritative DNS servers. 

Cache eviction operations are right now are only available via TCP control socket

```
dnsdist -e 'getPool(""):getCache():expungeByName(newDNSName("www.example.net"))'
```

Exposing the same functionality over HTTP RESTful APIs would ease the life-cycle management (via custom HTTP web-hooks) of the records cache without having to artificially lower its maximum TTL configuration into the `dnsdist.conf` file

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

Closes #6154 #10468.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/dnsdist-web.cc [hunk 1]
```diff
@@ -1458,6 +1463,91 @@ static void handleAllowFrom(const YaHTTP::Request& req, YaHTTP::Response& resp)
...
+    return;
+  }
+
+  DNSName name;
+  QType type(QType::ANY);
+  try {
+    name = DNSName(expungeName->second);
+  }
+  catch (const std::exception& e) {
+    resp.status = 404;
```
Comment (chbruyand) [inline pdns/dnsdist-web.cc:1504]:
```suggestion
    resp.status = 400;
```

File: pdns/dnsdist-web.cc [hunk 2]
```diff
@@ -350,6 +350,11 @@ static bool isMethodAllowed(const YaHTTP::Request& req)
       return true;
     }
   }
+  if (req.method == "DELETE") {
+    if (req.url.path == "/api/v1/cache") {
+      return true;
+    }
+  }
```
Comment (chbruyand) [inline pdns/dnsdist-web.cc:357]:
```suggestion
#ifndef DISABLE_WEB_CACHE_MANAGEMENT
  if (req.method == "DELETE") {
    if (req.url.path == "/api/v1/cache") {
      return true;
    }
  }
#endif /* DISABLE_WEB_CACHE_MANAGEMENT */
```

File: regression-tests.dnsdist/test_Caching.py [hunk 3]
```diff
@@ -2781,3 +2782,133 @@ def testCachingBackendSetRD(self):
...
+        self.assertEqual(r.status_code, 404)
+
+        # no 'pool'
+        r = requests.delete(url + '?name=cache-api.cache.tests.powerdns.com.&type=AAAA', headers=headers, timeout=self._webTimeout)
+        self.assertEqual(r.status_code, 400)
+
+        # no 'name'
+        r = requests.delete(url + '?pool=pool-without-cache&type=AAAA', headers=headers, timeout=self._webTimeout)
+        self.assertEqual(r.status_code, 400)
+
```
Comment (chbruyand) [inline regression-tests.dnsdist/test_Caching.py:2847]: Maybe add a test with an invalid name (too long label ?)

**Review by chbruyand — APPROVED** [review]
lgtm!

File: pdns/dnsdist-web.cc (hunk 1 above)
Comment (chbruyand) [reply pdns/dnsdist-web.cc:1504]: (this is just a personal taste)

Comment (rgacogne) [reply pdns/dnsdist-web.cc:1504]: You are 100% right, thanks!

File: regression-tests.dnsdist/test_Caching.py (hunk 3 above)
Comment (rgacogne) [reply regression-tests.dnsdist/test_Caching.py:2847]: Will do!

File: pdns/dnsdist-web.cc [hunk 4]
```diff
@@ -1458,6 +1465,91 @@ static void handleAllowFrom(const YaHTTP::Request& req, YaHTTP::Response& resp)
...
+    resp.body = Json(obj).dump();
+    return;
+  }
+
+  auto cache = pool->getCache();
+  if (cache == nullptr) {
+    resp.status = 404;
+    Json::object obj{
+      { "status", "not found" },
+      { "error", "there is no cache associated to the requested pool" },
```
[the remark this answers is not in the document]
Comment (rgacogne) [reply pdns/dnsdist-web.cc:1537]: Fixed in https://github.com/PowerDNS/pdns/pull/12500, thanks!

**Pull request merged**